enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,264 @@
+
+### Goals
+
+The MCO will take over management of the boot image references and the stub ignition. The installer is still responsible for creating the `MachineSet` at cluster bring-up of course, but once cluster installation is complete the MCO will ensure that boot images are in sync with the latest payload. From the user standpoint, this should cause less compatibility issues as nodes will no longer need to pivot to a different version of rhcos during node scaleup.
+
+This should not interfere with existing workflows such as Hive and ArgoCD. As this is an opt-in mechanism, the cluster admin will be protected against such scenarios of accidental "reconciliation". 
+
+### Non-Goals
+
+- The new subcontroller is only intended to support clusters that use MachineSet backed node scaling. This is meant to be a user opt-in feature, and if the user wishes to keep their boot images static it will let them do so.
+- This does not intend to solve [booting into custom pools](https://issues.redhat.com/browse/MCO-773). 
+- This does not target Hypershift, as [it does not use machinesets](https://github.com/openshift/hypershift/blob/32309b12ae6c5d4952357f4ad17519cf2424805a/hypershift-operator/controllers/nodepool/nodepool_controller.go#L2168).
+- This proposal only targets MAPI backed machinesets. It does not intend to support CAPI backed machinesets, but we hope to do so in a future release, perhaps with a seperate enhancement. 
+
+## Proposal
+
+__Overview__
+
+- The `machine-config-controller`(MCC) pod will gain a new sub-controller `machine_set_controller`(MSC) that monitors `MachineSet` changes and the `coreos-bootimages` [ConfigMap](https://github.com/openshift/installer/pull/4760) changes.
+- Before processing a MachineSet, the MSC will check if the following conditions are satisfied:
+  - `ManagedBootImages` feature gate is active
+  - The cluster and/or the machineset is opted-in to boot image updates. This mechanism is still TBD, see Workflow Description for more details.
+  - The golden configmap is verified to be in sync with the current version of the MCO. The MCO will "stamp"(annotate) the golden configmap with the new version of the MCO after atleast 1 node has succesfully completed an update to the new OCP image. This helps prevent `machinesets` being updated too soon at the end of a cluster upgrade, before the MCO itself has updated and has had a chance to roll out the new OCP image to the cluster. 
```

Comment:
Have you tested this? Will CVO overwrite this stamp since it is unaware of the annotation?

Reply:
So I believe the CVO will potentially overwrite the [configmap doing this](https://github.com/openshift/installer/pull/4760/files#diff-246420833c3e76c12ea9535984f811c8597f608a0504df7dd044a3d31bd629baR31-R35), losing the annotations. I had gated my check for this annotation to exist _and_ be a match, so we can account for those cases. I can test it to be doubly sure though.